### PR TITLE
[Modules] webhooks for ACR

### PR DESCRIPTION
**Pull request by tjav**
# Description

Adding webhooks to the Container registry module.
 
- Created new webhooks bicep template
- created markdown for module
- Added webhooks to main bicep file with defaults
- updated main markdown file

## Pipeline references
no pipeline run yet (please assist)

| Pipeline |
| - |
| [ContainerRegistry: Registries](https://github.com/Azure/ResourceModules/actions/workflows/ms.containerregistry.registries.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.ContainerRegistry/registries/deploy.bicep [hunk 1]
```diff
@@ -97,6 +97,9 @@ param zoneRedundancy string = 'Disabled'
 @description('Optional. All replications to create.')
 param replications array = []
 
+@description('Optional. All webhooks to create')
+param webhooks array = []
```
Comment (rahalan) [inline arm/Microsoft.ContainerRegistry/registries/deploy.bicep:101]: Please add a test case (add webhooks into a parameter file or create a new parameter file, that uses webhooks)

Comment (tjav) [reply arm/Microsoft.ContainerRegistry/registries/deploy.bicep:101]: check

Comment (rahalan) [reply arm/Microsoft.ContainerRegistry/registries/deploy.bicep:101]: Please let the pipeline run on your branch and add the successful pipeline run into the PR.

File: arm/Microsoft.ContainerRegistry/registries/readme.md [hunk 2]
```diff
@@ -10,15 +10,16 @@ Azure Container Registry is a managed, private Docker registry service based on
...
-| `Microsoft.ContainerRegistry/registries` | [2021-09-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.ContainerRegistry/2021-09-01/registries) |
-| `Microsoft.ContainerRegistry/registries/replications` | [2021-12-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.ContainerRegistry/2021-12-01-preview/registries/replications) |
-| `Microsoft.Insights/diagnosticSettings` | [2021-05-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Insights/2021-05-01-preview/diagnosticSettings) |
-| `Microsoft.Network/privateEndpoints` | [2021-05-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Network/2021-05-01/privateEndpoints) |
-| `Microsoft.Network/privateEndpoints/privateDnsZoneGroups` | [2021-02-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Network/2021-02-01/privateEndpoints/privateDnsZoneGroups) |
+| Resource Type                                             | API Version                                                                                                                                          |
```
Comment (eriqua) [inline arm/Microsoft.ContainerRegistry/registries/readme.md:13]: Please remove table autoformatting (you probably have a markdown VSC extension doing that)

File: arm/Microsoft.ContainerRegistry/registries/webhooks/README.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Container Registry Webhooks
```
Comment (eriqua) [inline arm/Microsoft.ContainerRegistry/registries/webhooks/README.md:1]: Please remove table autoformatting as per the other readme, plus please rename the file to lowercase `readme.md`

File: arm/Microsoft.ContainerRegistry/registries/readme.md (hunk 2 above)
Comment (MariusStorhaug) [reply arm/Microsoft.ContainerRegistry/registries/readme.md:13]: If you are looking for the setting to disable:

https://github.com/Azure/ResourceModules/blob/11ecea725adfa6a585de12344389eb0c4e9876f5/.vscode/settings.json#L24

File: arm/Microsoft.ContainerRegistry/registries/webhooks/README.md (hunk 3 above)
Comment (MariusStorhaug) [reply arm/Microsoft.ContainerRegistry/registries/webhooks/README.md:1]: If you are looking for the setting to disable:

https://github.com/Azure/ResourceModules/blob/11ecea725adfa6a585de12344389eb0c4e9876f5/.vscode/settings.json#L24

File: arm/Microsoft.ContainerRegistry/registries/readme.md [hunk 4]
```diff
@@ -10,15 +10,16 @@ Azure Container Registry is a managed, private Docker registry service based on
...
+| `Microsoft.ContainerRegistry/registries/replications`     | [2021-12-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.ContainerRegistry/2021-12-01-preview/registries/replications)        |
+| `Microsoft.ContainerRegistry/registries/webhooks`         | [2021-12-01-preview](https://docs.microsoft.com/en-us/azure/templates/microsoft.containerregistry/2021-12-01-preview/registries/webhooks?tabs=bicep) |
+| `Microsoft.Insights/diagnosticSettings`                   | [2021-05-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Insights/2021-05-01-preview/diagnosticSettings)                      |
+| `Microsoft.Network/privateEndpoints`                      | [2021-05-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Network/2021-05-01/privateEndpoints)                                         |
+| `Microsoft.Network/privateEndpoints/privateDnsZoneGroups` | [2021-02-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Network/2021-02-01/privateEndpoints/privateDnsZoneGroups)                    |
```
Comment (MariusStorhaug) [inline arm/Microsoft.ContainerRegistry/registries/readme.md:22]:
```suggestion
| Resource Type | API Version |
| :-- | :-- |
| `Microsoft.Authorization/locks` | [2017-04-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2017-04-01/locks) |
| `Microsoft.Authorization/roleAssignments` | [2020-10-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2020-10-01-preview/roleAssignments) |
| `Microsoft.ContainerRegistry/registries` | [2021-09-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.ContainerRegistry/2021-09-01/registries) |
| `Microsoft.ContainerRegistry/registries/replications` | [2021-12-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.ContainerRegistry/2021-12-01-preview/registries/replications) |
| `Microsoft.ContainerRegistry/registries/webhooks` | [2021-12-01-preview](https://docs.microsoft.com/en-us/azure/templates/microsoft.containerregistry/2021-12-01-preview/registries/webhooks?tabs=bicep) |
| `Microsoft.Insights/diagnosticSettings` | [2021-05-01-preview](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Insights/2021-05-01-preview/diagnosticSettings) |
| `Microsoft.Network/privateEndpoints` | [2021-05-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Network/2021-05-01/privateEndpoints) |
| `Microsoft.Network/privateEndpoints/privateDnsZoneGroups` | [2021-02-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Network/2021-02-01/privateEndpoints/privateDnsZoneGroups) |
```

File: arm/Microsoft.ContainerRegistry/registries/readme.md [hunk 5]
```diff
@@ -184,11 +185,11 @@ You can specify multiple user assigned identities to a resource by providing add
...
-| `name` | string | The Name of the Azure container registry. |
-| `resourceGroupName` | string | The name of the Azure container registry. |
-| `resourceId` | string | The resource ID of the Azure container registry. |
+| Output Name                 | Type   | Description                                       |
+| :-------------------------- | :----- | :------------------------------------------------ |
+| `location`                  | string | The location the resource was deployed into.      |
+| `loginServer`               | string | The reference to the Azure container registry.    |
+| `name`                      | string | The Name of the Azure container registry.         |
+| `resourceGroupName`         | string | The name of the Azure container registry.         |
+| `resourceId`                | string | The resource ID of the Azure container registry.  |
```
Comment (MariusStorhaug) [inline arm/Microsoft.ContainerRegistry/registries/readme.md:194]:
```suggestion
| Output Name | Type | Description |
| :-- | :-- | :-- |
| `location` | string | The location the resource was deployed into. |
| `loginServer` | string | The reference to the Azure container registry. |
| `name` | string | The Name of the Azure container registry. |
| `resourceGroupName` | string | The name of the Azure container registry. |
| `resourceId` | string | The resource ID of the Azure container registry. |
```

File: arm/Microsoft.ContainerRegistry/registries/deploy.bicep [hunk 6]
```diff
@@ -247,6 +250,27 @@ module registry_replications 'replications/deploy.bicep' = [for (replication, in
   }
 }]
 
+module registry_webhooks 'webhooks/deploy.bicep' = [for (webhook, index) in webhooks: {
+  name: '${uniqueString(deployment().name, location)}-Registry-Replication-${index}'
```
Comment (MariusStorhaug) [inline arm/Microsoft.ContainerRegistry/registries/deploy.bicep:254]:
```suggestion
  name: '${uniqueString(deployment().name, location)}-Registry-Webhooks-${index}'
```

File: arm/Microsoft.ContainerRegistry/registries/webhooks/README.md (hunk 3 above)
Comment (MariusStorhaug) [reply arm/Microsoft.ContainerRegistry/registries/webhooks/README.md:1]: This its more about generating the readme with the script. I'll fix that so we can progress on the PR.

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]

File: arm/Microsoft.ContainerRegistry/registries/readme.md (hunk 2 above)
Comment (MariusStorhaug) [reply arm/Microsoft.ContainerRegistry/registries/readme.md]: Fixed

File: arm/Microsoft.ContainerRegistry/registries/webhooks/README.md (hunk 3 above)
Comment (MariusStorhaug) [reply arm/Microsoft.ContainerRegistry/registries/webhooks/README.md]: Regenerated readme with new script

**Review by MariusStorhaug — APPROVED** [review]

**Pull request merged**